File: lc0019.go
// LeetCode problem 0023: Merge K Sorted Lists
// https://leetcode.com/problems/merge-k-sorted-lists/description/

package lc0019

import (
	"fmt"
)

type ListNode struct {
	Val  int32
	Next *ListNode
}

func (l *ListNode) ToString() string {
	var buf string
	buf += fmt.Sprintf("ListNode { Val = %v, Next = ", l.Val)
	if l.Next != nil {
		buf += l.Next.ToString()
	} else {
		buf += "nil }\n"
	}

	return buf
}

func main() {
	list := ListNode{Val: 1, Next: &ListNode{Val: 2, Next: &ListNode{Val: 3, Next: &ListNode{Val: 4, Next: &ListNode{Val: 5}}}}}
	result := removeNthFromEnd(&list, 5)
	fmt.Printf("%v\n", result.ToString())
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	length := 0
	current := head
	for current != nil {
		current = current.Next
		length++
	}

	if length == n {
		return head.Next
	}

	h := head
	current = h
	for i := 0; i < length; i++ {
		if i == length-n-1 {
			current.Next = current.Next.Next
			break
		} else {
			current = current.Next
		}
	}

	return h
}

File: lc0019.rs
// LeetCode problem 0019: Remove Nth Node From End of List
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/description/

#[derive(PartialEq, Eq, Clone, Debug)]
struct ListNode {
    val: i32,
    next: Option<Box<ListNode>>
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode {
            next: None,
            val
        }
    }

    fn build_list(vals: Vec<i32>) -> Option<Box<ListNode>> {
        if vals.len() == 0 {
            return None;
        }

        let mut node = ListNode::new(vals[0]);
        node.next = Self::build_list(vals[1..].to_vec());

        return Some(Box::new(node));
    }
}

fn main() {
    let result = remove_nth_from_end(ListNode::build_list(vec![1,2,3,4,5]), 2);
    println!("Result = {result:?}");
}

fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {
    let mut len = 0;
    let mut current = head.as_ref();
    while let Some(n) = current {
        len += 1;
        current = n.next.as_ref();
    }

    println!("Len = {len}");
    println!("Head is {head:?}");
    let start = head.as_ref();
    current = head.as_ref();
    for i in 0..len {
        if i == len - n {
            current.as_mut().unwrap().next = Some(current.unwrap().next.clone().unwrap().next.unwrap());
            break;
        }
        current = current.unwrap().next.as_ref();
    }

    return start.cloned();
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn example_one() {
        let result = remove_nth_from_end(ListNode::build_list(vec![1,2,3,4,5]), 2);
        assert_eq!(result, ListNode::build_list(vec![1,2,3,5]));
    }

    #[test]
    fn example_two() {
        let result = remove_nth_from_end(ListNode::build_list(vec![1]), 1);
        assert_eq!(result, ListNode::build_list(vec![]));
    }
    
    #[test]
    fn example_one() {
        let result = remove_nth_from_end(ListNode::build_list(vec![1,2]), 1);
        assert_eq!(result, ListNode::build_list(vec![1]));
    }
}
